File: money.py
"""
money.py
--------
Amounts as whole numbers of cents.

Deliberately much smaller than the wallet app's equivalent. That one has to
read a figure out of any bank's CSV, so it carries rules for both decimal
conventions and for accounting negatives. Nothing here reads a bank file:
every amount arrives from a keypad in this app's own page, so it needs one
parser and one formatter, and inventing the rest would be speculative.

What it keeps is the part that matters. Money is an integer here because
floats are the wrong type for it: 0.1 + 0.2 is 0.30000000000000004, so a
day's total disagrees with the entries above it, and a reader who notices
that stops trusting every other figure on the screen.
"""
import re

# The currencies this app offers. Euro first because that is where the
# spending happens; the other two are here so a trip does not need a second
# app. Which one an entry is in is stored per entry -- see db.py on why a
# tracker must not convert on the way in.
CURRENCIES = ("EUR", "CAD", "USD")
DEFAULT_CURRENCY = "EUR"

SYMBOLS = {"EUR": "€", "CAD": "CA$", "USD": "US$"}

# Two decimal places for all three. Named rather than written as 100 so the
# assumption is visible: JPY has none, and this is the line that would have to
# change.
MINOR_UNITS = 2
_SCALE = 10 ** MINOR_UNITS

_ALLOWED = re.compile(r"[^\d.,]")


class MoneyError(ValueError):
    """An amount that could not be read.

    Never a silent zero. An entry of nothing looks exactly like an entry
    somebody meant to make, so it has to be refused at the edge rather than
    stored and puzzled over later.
    """


def parse(text):
    """A typed amount as an integer number of cents.

    One convention, not two: the keypad in this app produces "12.34", and a
    comma is accepted as the same separator because a phone keyboard set to a
    European locale offers one. There is no thousands-grouping rule, because
    nothing here is typed with grouping -- which is the difference between
    this and the wallet app's parser, and the reason that one is four times
    the size.
    """
    if text is None:
        raise MoneyError("no amount given")
    if isinstance(text, int) and not isinstance(text, bool):
        return text * _SCALE

    raw = str(text)
    # A sign has to be caught before the strip, because the strip removes it:
    # "-5" became 5, so typing a minus recorded a five-euro expense instead of
    # refusing. Silently reversing what somebody typed is worse than either
    # accepting or rejecting it, and this app records only what was spent.
    if "-" in raw:
        raise MoneyError("an expense is a positive amount")

    cleaned = _ALLOWED.sub("", raw).replace(",", ".")
    if not any(character.isdigit() for character in cleaned):
        raise MoneyError(f"no digits in {text!r}")
    if cleaned.count(".") > 1:
        raise MoneyError(f"cannot read {text!r}")

    whole, _, fraction = cleaned.partition(".")
    if len(fraction) > MINOR_UNITS:
        raise MoneyError(f"{text!r} has more than {MINOR_UNITS} decimal places")

    cents = int(whole or 0) * _SCALE + int((fraction + "00")[:MINOR_UNITS])
    if cents <= 0:
        raise MoneyError("an amount has to be more than nothing")
    return cents


def format(cents, currency=DEFAULT_CURRENCY):
    """For the screen: a symbol, grouping, always two decimals.

    Takes a number, not None. It had a `symbol=False` flag and a
    `cents is None` guard, and neither had a caller: `plain` is the
    without-a-symbol formatter, and no column or query in this app yields a
    null amount -- `amount` is NOT NULL, every grouped SUM has at least one
    row by definition, and the two ungrouped sums COALESCE in SQL. Returning
    "" for None only meant a blank where a figure belongs, which is a harder
    failure to notice than the TypeError this raises now.
    """
    sign = "-" if cents < 0 else ""
    whole, fraction = divmod(abs(int(cents)), _SCALE)
    return (f"{sign}{SYMBOLS.get(currency, currency + ' ')}"
            f"{whole:,}.{fraction:0{MINOR_UNITS}d}")


def plain(cents):
    """The bare figure for a CSV: no symbol and no grouping.

    "1,234.56" is two fields to anything that splits on commas, which is the
    one thing a CSV reader reliably does.
    """
    sign = "-" if cents < 0 else ""
    whole, fraction = divmod(abs(int(cents)), _SCALE)
    return f"{sign}{whole}.{fraction:0{MINOR_UNITS}d}"


def total(amounts):
    """Sum of cents. Exists so no caller is tempted to sum floats."""
    return sum(int(a) for a in amounts if a is not None)


def known(currency):
    """The currency, upper-cased, or the default.

    Falls back rather than raising: a stale tab posting a currency this
    version no longer offers should still record the expense, because losing
    the entry is worse than filing it in euros and letting it be corrected.
    """
    code = (currency or "").strip().upper()
    return code if code in CURRENCIES else DEFAULT_CURRENCY

File: docs/app/js/money.js
/* ---------------------------------------------------------------------------
   money.js — money.py, in the browser
   ---------------------------------------------------------------------------
   A line-for-line port of money.py, and the one file in this bundle that is
   not allowed to be "close enough". Everything else here is presentation; this
   decides what a number *is*.

   money.py's whole argument is that an amount is an integer number of cents,
   because 0.1 + 0.2 is 0.30000000000000004 and a day's total that disagrees
   with the rows printed above it costs a reader their trust in every other
   figure on the screen. A port that quietly reintroduces a float, or rounds
   the third decimal place instead of refusing it, would put that back — and
   it would do it on the *published* build, which is the copy most people see.

   So this is not trusted. `tools/build_static.py` runs this exact file in a
   real browser against the real money.py over several hundred values —
   parsing, formatting, the plain CSV form, totals and currency fallback — and
   refuses to write docs/app/ if any one of them disagrees, down to the text of
   the error message.

   Four places where JavaScript cannot be made to match. None of them is
   reachable from the keypad, all of them are narrowed rather than papered
   over, and each is named in build_static.py's case list as an exclusion, so
   the gap between "checked" and "identical" is written down in both files
   rather than in neither:

     * Python has int and float; JavaScript has one number type. money.py
       short-circuits `isinstance(text, int)` to `text * 100`, so `parse(0)`
       returns 0 where `parse(0.0)` goes through the text path and raises. This
       file cannot tell those apart and takes the int branch for any integral
       number. The build's case list therefore holds no integral floats.
     * Python integers are unbounded. Above 2**53 cents — ninety trillion euro
       — this loses precision and money.py does not. The keypad stops at
       99999999 cents, so nothing in the app can reach it, but it is a
       difference and not a rounding.
     * Python's `\d` matches every Unicode decimal digit; JavaScript's matches
       0-9. So money.py reads an Arabic-Indic '٠' as a zero and this file
       strips it as punctuation. Both refuse the amount; they disagree about
       which message to refuse it with. Making the port agree would mean
       carrying a Unicode table to parse digits nothing in this app can type.
     * `repr()` below is Python's repr for the printable cases only. Python
       escapes an unprintable character as `\xa0`; this prints it. Only error
       messages are affected, and only for input the app cannot produce.
   --------------------------------------------------------------------------- */

var TallyMoney = (function () {
  'use strict';

  /* Same order as money.py: euro first, because that is where the spending
     happens. */
  var CURRENCIES = ['EUR', 'CAD', 'USD'];
  var DEFAULT_CURRENCY = 'EUR';

  var SYMBOLS = { EUR: '€', CAD: 'CA$', USD: 'US$' };

  var MINOR_UNITS = 2;
  var SCALE = Math.pow(10, MINOR_UNITS);

  var ALLOWED = /[^\d.,]/g;

  /* money.py raises MoneyError, a ValueError subclass. The shim in
     static-api.js turns one into a 400 with the same body the Flask app
     sends, so the page's error handling does not need to know which build it
     is running in. A tagged Error rather than a subclass: the bundle targets
     browsers old enough that `class X extends Error` breaks instanceof. */
  function MoneyError(message) {
    var error = new Error(message);
    error.name = 'MoneyError';
    return error;
  }

  /* Python's repr(), for the handful of types that reach an error message.
     The messages are part of what the build compares -- an error a person
     reads is as much the behaviour as the number is -- and they interpolate
     `{text!r}`, so matching them means matching repr. */
  function repr(value) {
    if (value === null || value === undefined) return 'None';
    if (value === true) return 'True';
    if (value === false) return 'False';
    if (typeof value === 'number') return String(value);
    var text = String(value);
    // Python prefers single quotes, and switches to double only when the
    // string holds a single quote and no double quote.
    var quote = (text.indexOf("'") >= 0 && text.indexOf('"') < 0) ? '"' : "'";
    var out = text.replace(/\\/g, '\\\\');
    if (quote === "'") out = out.replace(/'/g, "\\'");
    return quote + out.replace(/\n/g, '\\n').replace(/\r/g, '\\r')
                      .replace(/\t/g, '\\t') + quote;
  }

  function group(whole) {
    /* Python's `{n:,}`. */
    return String(whole).replace(/\B(?=(\d{3})+(?!\d))/g, ',');
  }

  function pad(fraction) {
    var text = String(fraction);
    while (text.length < MINOR_UNITS) text = '0' + text;
    return text;
  }

  function parse(text) {
    if (text === null || text === undefined) throw MoneyError('no amount given');
    /* `isinstance(text, int) and not isinstance(text, bool)`. Booleans fall
       through to the text path in money.py and fail there for having no
       digits, so they must fall through here too. */
    if (typeof text === 'number' && isFinite(text) && Math.floor(text) === text) {
      return text * SCALE;
    }

    var raw = String(text);
    /* Caught before the strip, because the strip removes it: "-5" became 5,
       so typing a minus recorded a five-euro expense instead of refusing. */
    if (raw.indexOf('-') >= 0) throw MoneyError('an expense is a positive amount');

    var cleaned = raw.replace(ALLOWED, '').replace(/,/g, '.');
    if (!/\d/.test(cleaned)) throw MoneyError('no digits in ' + repr(text));
    if (cleaned.split('.').length - 1 > 1) throw MoneyError('cannot read ' + repr(text));

    var cut = cleaned.indexOf('.');
    var whole = cut < 0 ? cleaned : cleaned.slice(0, cut);
    var fraction = cut < 0 ? '' : cleaned.slice(cut + 1);
    if (fraction.length > MINOR_UNITS) {
      throw MoneyError(repr(text) + ' has more than ' + MINOR_UNITS +
                       ' decimal places');
    }

    var cents = Number(whole || 0) * SCALE + Number((fraction + '00').slice(0, MINOR_UNITS));
    if (cents <= 0) throw MoneyError('an amount has to be more than nothing');
    return cents;
  }

  function format(cents, currency) {
    if (currency === undefined) currency = DEFAULT_CURRENCY;
    /* money.py takes a number, not None, and raises a TypeError rather than
       returning "" -- a blank where a figure belongs is harder to notice than
       a crash. */
    if (cents === null || cents === undefined || typeof cents !== 'number') {
      throw new TypeError('format() needs a number of cents, got ' + repr(cents));
    }
    var sign = cents < 0 ? '-' : '';
    var size = Math.abs(Math.trunc(cents));
    /* hasOwnProperty, not `SYMBOLS[currency] ||`: a currency named
       "constructor" or "toString" would otherwise find something on the
       prototype and print it as the symbol. */
    var symbol = Object.prototype.hasOwnProperty.call(SYMBOLS, currency)
      ? SYMBOLS[currency] : currency + ' ';
    return sign + symbol + group(Math.floor(size / SCALE)) + '.' + pad(size % SCALE);
  }

  function plain(cents) {
    /* No symbol and no grouping: "1,234.56" is two fields to anything that
       splits on commas, which is the one thing a CSV reader reliably does. */
    var sign = cents < 0 ? '-' : '';
    var size = Math.abs(Math.trunc(cents));
    return sign + String(Math.floor(size / SCALE)) + '.' + pad(size % SCALE);
  }

  function total(amounts) {
    var sum = 0;
    for (var i = 0; i < amounts.length; i++) {
      if (amounts[i] === null || amounts[i] === undefined) continue;
      sum += Math.trunc(Number(amounts[i]));
    }
    return sum;
  }

  function known(currency) {
    /* Falls back rather than raising: losing the entry is worse than filing
       it in euros and letting it be corrected. */
    var code = String(currency ? currency : '').trim().toUpperCase();
    return CURRENCIES.indexOf(code) >= 0 ? code : DEFAULT_CURRENCY;
  }

  return {
    CURRENCIES: CURRENCIES,
    DEFAULT_CURRENCY: DEFAULT_CURRENCY,
    SYMBOLS: SYMBOLS,
    MINOR_UNITS: MINOR_UNITS,
    MoneyError: MoneyError,
    repr: repr,
    parse: parse,
    format: format,
    plain: plain,
    total: total,
    known: known
  };
})();

if (typeof module !== 'undefined' && module.exports) module.exports = TallyMoney;
